sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -545,29 +545,73 @@ ssize_t coap_tree_handler(coap_pkt_t *pkt, uint8_t *resp_buf, unsigned resp_buf_
     return coap_build_reply(pkt, COAP_CODE_404, resp_buf, resp_buf_len, 0);
 }
 
+ssize_t coap_build_reply_header(coap_pkt_t *pkt, unsigned code,
+                                void *buf, size_t len,
+                                int ct,
+                                void **payload, size_t *payload_len_max)
+{
+    uint8_t *bufpos = buf;
+    uint32_t no_response;
+    uint8_t type = coap_get_type(pkt) == COAP_TYPE_CON
+                 ? COAP_TYPE_ACK
+                 : COAP_TYPE_NON;
+
+    bufpos += coap_build_hdr(buf, type, coap_get_token(pkt), coap_get_token_len(pkt),
+                             code, ntohs(pkt->hdr->id));
+
+    if (coap_opt_get_uint(pkt, COAP_OPT_NO_RESPONSE, &no_response) == 0) {
+        const uint8_t no_response_index = (code >> 5) - 1;
+        /* If the handler code misbehaved here, we'd face UB otherwise */
+        assume(no_response_index < 7);
+
+        const uint8_t mask = 1 << no_response_index;
+        if (no_response & mask) {
+            if (payload) {
+                *payload = NULL;
+                *payload_len_max = 0;
+                payload = NULL;
+            }
+            if (type != COAP_TYPE_ACK) {
+                return 0;
```

Comment:
According to the doc, the function should return the size of the header written to buf.

Reply:
That's intentional: If the no-response option is set, the client does not want any answer (save for the ACK of a confirmable message), so if we wouldn't send an ACK, don't send anything at all by returning 0 here.

I added a comment.

Reply:
Humm I see, but it still clashes with the documentation. If you really want to signal this special condition by returning a zero, it should be mentioned in the docs, I'd say.

Reply:
I add it, but IMHO this still fits under `size of reply header on success` - the size is just 0